**Importing the Libraries**

In [1]:
%load_ext autoreload
%autoreload 2
%matplotlib inline

from fastai.structured import *
from fastai.imports import *

from pandas_summary import DataFrameSummary
from sklearn.ensemble import RandomForestRegressor, RandomForestClassifier
from IPython.display import display
from sklearn import metrics

**Reading the Dataset**

In [2]:
df_raw = pd.read_csv('../input/train.csv')
df_raw.head()

ID       y  X0 X1  X2 X3 X4  ...  X378 X379 X380  X382  X383  X384  X385
0   0  130.81   k  v  at  a  d  ...     0    0    0     0     0     0     0
1   6   88.53   k  t  av  e  d  ...     0    0    0     0     0     0     0
2   7   76.26  az  w   n  c  d  ...     0    0    0     1     0     0     0
3   9   80.62  az  t   n  f  d  ...     0    0    0     0     0     0     0
4  13   78.02  az  v   n  f  d  ...     0    0    0     0     0     0     0

[5 rows x 378 columns]

**Competition Details:**
*In this competition, Daimler is challenging Kagglers to tackle the curse of dimensionality and reduce the time that cars spend on the test bench. Competitors will work with a dataset representing different permutations of Mercedes-Benz car features to predict the time it takes to pass testing. Winning algorithms will contribute to speedier testing, resulting in lower carbon dioxide emissions without reducing Daimler’s standards.*

In [3]:
train_cats(df_raw)

*Using the : **train_cats(dataframe)** to Change any columns of strings in a panda's dataframe to a column of categorical values. This applies the changes inplace.*

*Checking for Missing Values*

In [4]:
(df_raw.isnull().sum().sort_index()/len(df_raw)).sum()

0.0

**No missing values in the dataset**

In [6]:
df, y, nas = proc_df(df_raw, 'y')

*Now we have separated **y** column from df_raw and returned dataframe into df, and **y** (Target Variable) into y. Along with handling categorical data.*

In [7]:
df.head()

ID  X0  X1  X2  X3  X4  X5  ...   X378  X379  X380  X382  X383  X384  X385
0   0  33  24  18   1   4  25  ...      0     0     0     0     0     0     0
1   6  33  22  20   5   4  29  ...      0     0     0     0     0     0     0
2   7  21  25  35   3   4  28  ...      0     0     0     1     0     0     0
3   9  21  22  35   6   4  28  ...      0     0     0     0     0     0     0
4  13  21  24  35   6   4  13  ...      0     0     0     0     0     0     0

[5 rows x 377 columns]

**Lets Build our First Model using Random Forest with 10 estimators**

In [8]:
m = RandomForestRegressor(n_jobs=-1)
m.fit(df, y)
m.score(df,y)

/opt/conda/lib/python3.6/site-packages/sklearn/ensemble/forest.py:248: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


0.9144472140467663

*Lets split the df and y into train sets and validation sets to check whether model overfits on validation set*

In [9]:
def split_vals(a,n): return a[:n].copy(), a[n:].copy()
n_valid = 800  # same as Kaggle's test set size
n_trn = len(df)-n_valid
raw_train, raw_valid = split_vals(df_raw, n_trn)
X_train, X_valid = split_vals(df, n_trn)
y_train, y_valid = split_vals(y, n_trn)
X_train.shape, y_train.shape, X_valid.shape

((3409, 377), (3409,), (800, 377))

In [10]:
df.shape

(4209, 377)

**print_score(model)**:  *Function to print r^2 score on training set and validation set*

In [11]:
def print_score(m):
    res = [m.score(X_train, y_train), m.score(X_valid, y_valid)]
    if hasattr(m, 'oob_score_'): res.append(m.oob_score_)
    print(res)

In [12]:
set_rf_samples(2600)

**Lets Tune the Hyper parameters of our Model**

In [18]:
m = RandomForestRegressor(n_estimators=100,n_jobs=-1)
%time m.fit(X_train, y_train)
print_score(m)

CPU times: user 12.6 s, sys: 48 ms, total: 12.6 s
Wall time: 3.26 s
[0.8969674581490968, 0.5381520360559964]


**We got R^2 Score of:**
**0.89 on training set**
**0.53 on Validation set**

In [19]:
set_rf_samples(2800)

In [20]:
m = RandomForestRegressor(n_estimators=200,n_jobs=-1,min_samples_leaf=5)
%time m.fit(X_train, y_train)
print_score(m)

CPU times: user 12.9 s, sys: 28 ms, total: 12.9 s
Wall time: 3.4 s
[0.7076122180784353, 0.6112338244455511]


**Our Model is now not too much over fitting on the validation set.**

In [22]:
m = RandomForestRegressor(n_estimators=200,n_jobs=-1,min_samples_leaf=10)
%time m.fit(X_train, y_train)
print_score(m)

CPU times: user 11.1 s, sys: 24 ms, total: 11.1 s
Wall time: 2.9 s
[0.6442950811144379, 0.6333180933749857]


**Our Model is now not overfitting  much**


Coefficient of Determination (R^2) Score are:
1. Training Set: 0.644*
2. Validation Set: 0.633

*Making Predictions and submission to Kaggle on the test set*

In [23]:
df_raw_test = pd.read_csv('../input/test.csv')

In [24]:
df_raw_test.head()

ID  X0 X1  X2 X3 X4 X5  ...  X378 X379  X380  X382  X383  X384  X385
0   1  az  v   n  f  d  t  ...     1    0     0     0     0     0     0
1   2   t  b  ai  a  d  b  ...     0    0     0     0     0     0     0
2   3  az  v  as  f  d  a  ...     1    0     0     0     0     0     0
3   4  az  l   n  f  d  z  ...     1    0     0     0     0     0     0
4   5   w  s  as  c  d  y  ...     0    0     0     0     0     0     0

[5 rows x 377 columns]

In [25]:
apply_cats(df_raw_test,df_raw)

In [26]:
df_raw_test['y']=df_raw['y']

In [27]:
dftest, ytest, nastest = proc_df(df_raw_test, 'y')

In [28]:
ypred = m.predict(dftest)

In [29]:
sub = pd.DataFrame()

In [30]:
sub['ID']=dftest['ID']
sub['y']=ypred

In [31]:
sub.to_csv('solution1.csv',index=False)

In [32]:
sub.head()

ID           y
0   1   78.749070
1   2   94.798175
2   3   78.144532
3   4   78.708017
4   5  115.153716

1. Our Model got coefficient of determination of 0.55162**
2. Winning Model got coefficient of determination of 0.5555**